enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,172 @@
+- Producers - each CLI that wants to offer its binaries in-cluster is responsible for providing its metadata and description via a ClusterCLI as well as
+  a CLI image to deliver its artifacts. 
+
+### Non-Goals
+
+* CLI manager will not build or serve the binaries.  It will know where to find them.
+* This proposal is not concerned with _which_ binaries will be managed.  This proposal is meant to determine the mechanism only.  Consumers and publishers are clients of the mechanism. 
+* CLI manager will not create or update the ClusterCLI Custom Resources, those will be managed by individual CLIs and their operators. 
+
+The ClusterCLI CRD will be installed either by a new lightweight operator or by an existing operator.
+The ClusterCLI/oc Custom Resource will serve as a reference consumer implementation for other CLIs to follow.    
+
+[for an example I've created a ClusterCLI controller that will populate 3 test CRs, oc, odo, and kn](https://github.com/sallyom/clustercli)
+
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user, I want a CLI in-cluster manager for various CLIs available for OpenShift and related services so that I can discover, install and upgrade them
+
+##### example:
+```
+$ oc krew list
+NAME                 DESC                        LATEST       INSTALLED
+-----                -----                       -----        -----
+openshift-installer  OpenShift Installer CLI     4.4          4.3
+kubectl              Kubernetes CLI              1.15         1.13
+oc                   OpenShift CLI               4.4          4.3
+odo                  OpenShift Developer CLI     1.0          Not Installed
+kn                   Knative CLI                 1.0          Not Installed
+tkn                  Tekton CLI                  1.0          Not Installed
+
+
+$ oc krew install odo
+$ oc krew install kn
+$ oc krew update oc
+$ oc krew remove kn
+```
+
+## References
+- [Krew: kubectl plugin manager](https://github.com/GoogleContainerTools/krew/) - manages kubectl plugins from 
+an [index](https://github.com/kubernetes-sigs/krew-index) of all known krew plugins.
+    
+- ["Uc" PoC by Hiram](https://github.com/chirino/uc) - manages Kubernetes CLI clients with an online catalog of releases.  Installs to a user's
+home directory, $HOME/.uc/cache  and when the cluster version does not match a known version, will install latest (well 'latest' known for uc oc
+atm is 3.11)
+
+_notes on macOS binaries_       
+[Signing binaries for macOS Catalina](https://developer.apple.com/news/?id=09032019a)    
+[related to above, Go toolchain issue with macOS Catalina](https://github.com/golang/go/issues/34986)    
+
+## Infrastructure Needed
+
+- controller/operator - for installing ClusterCLI CRD and creating in-cluster Krew index
+- new CI repository onboarded
+- extend ConsoleCLIDownload CRD and API for cli-manager [example](https://github.com/sallyom/api/blob/add-clustercli/config/v1/0000_03_config-operator_01_clustercli.crd.yaml)
```

Comment:
We need a new CRD entirely, the one from console is not much of use for us.

Reply:
ok, updated